src/network/networkd-manager-varlink.c

```diff
@@ -185,12 +185,20 @@ static int vl_method_set_persistent_storage(Varlink *vlink, JsonVariant *paramet
                 return r;
 
         if (ready) {
-                _cleanup_close_ int fd = -EBADF;
                 struct stat st, st_prev;
+                int fd;
 
-                fd = open("/var/lib/systemd/network/", O_CLOEXEC | O_DIRECTORY | O_PATH);
+                fd = varlink_peek_fd(vlink, 0);
                 if (fd < 0)
-                        return log_warning_errno(errno, "Failed to open /var/lib/systemd/network/: %m");
+                        return log_warning_errno(fd, "Failed to peek file descriptor of the persistent storage: %m");
+
+                r = fd_verify_safe_flags_full(fd, O_DIRECTORY);
+                if (r == -EREMOTEIO)
+                        return log_warning_errno(r, "Passed persistent storage fd has unexpected flags.");
+                if (r < 0)
+                        return log_warning_errno(r, "Failed to verify flags of passed persistent storage fd: %m");
+                if (!FLAGS_SET(r, O_DIRECTORY))
+                        return log_warning_errno(SYNTHETIC_ERRNO(EINVAL), "Passed persistent storage is not a directory: %m");
```

Comment:
`stat_verify_directory` is used below, and I don't think such redundancy is needed? The former also gives more specific errno.

BTW, I'm not sure if the kernel would return `O_DIRECTORY` if that's not explicitly specified when the fd is opened. So better not take this too seriously anyway?